### ISSUE TITLE
Default failure message for custom matcher incorrectly formats `expected` hashes

### ISSUE THREAD

**Issue by reshleman** [rspec/rspec-expectations#1192]
### Subject of the issue

When using a hash as the `expected` value in a custom matcher, the default failure message formats the `expected` value as a list of 2-element arrays.

(An `actual` value which is a hash is formatted as expected in the default error message.)

### Your environment
* Ruby version: 2.6.5
* rspec-expectations version: 3.9.2

### Steps to reproduce
```ruby
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  gem "rspec", "3.9.0" # Activate the gem and version you are reporting the issue against.
end

puts "Ruby version is: #{RUBY_VERSION}"
require 'rspec/autorun'

RSpec::Matchers.define :custom_match do |expected|
  match do |actual|
    actual == expected
  end
end

# These are all intended to fail in order to demonstrate the failure messages
# for different `expected` values.
RSpec.describe 'default failure message' do
  it 'correctly formats a string' do
    expect(nil).to custom_match('foo')
  end

  it 'correctly formats an Object' do
    expect(nil).to custom_match(Object.new)
  end

  it 'correctly formats an array' do
    expect(nil).to custom_match([1, 'foo', Object.new])
  end

  it 'incorrectly formats a hash as an array' do
    expect(nil).to custom_match({a: 1, foo: 'bar', bang: Object.new})
  end
end
```

### Expected behavior
When a hash is the `expected` object, the failure message should format the hash as a hash.

### Actual behavior
When a hash is the `expected` object, the failure message formats the hash as a list of 2-element key-value arrays:

```
  1) default failure message incorrectly formats a hash as an array
     Failure/Error: expect(nil).to custom_match({a: 1, foo: 'bar', bang: Object.new})
       expected nil to custom match [:a, 1], [:foo, "bar"], and [:bang, #<Object:0x00007fecf73d6dc8>]
     # rspec_report.rb:39:in `block (2 levels) in <main>'
```

---

I believe this is a result of the use of `EnglishPhrasing.list` in `BaseMatcher#description`:

https://github.com/rspec/rspec-expectations/blob/eb1787f0e1a5ec2c2650048ee60a45d4c9738d78/lib/rspec/matchers/built_in/base_matcher.rb#L60-L64

`EnglishPhrasing.list` in turn appears to format all enumerables, including hashes, as a list:

https://github.com/rspec/rspec-expectations/blob/eb1787f0e1a5ec2c2650048ee60a45d4c9738d78/lib/rspec/matchers/english_phrasing.rb#L26-L39

I'm happy to open a PR to fix, provided some guidance on the preferred solution.

**Comment by pirj** [comment]
First of all - great report, thank you!

The culprit seems to be here:
```ruby
Array({a: 1, b: 2})
=> [[:a, 1], [:b, 2]]
```

For hashes, this representation isn't the best one possible, so maybe it deserves a special case in the `self.list`.

You may start with adding a check if it's a Hash, and return some abracadabra to see if any of the specs/features fail. If no - we don't have this behaviour defined, and you're mostly free to sculpt it the way you see fit.

**Comment by reshleman** [comment]
Thank you for the speedy response! I took a quick pass at a fix in #1193.
